_posts/2020-02-17-constructor-injection.md

```diff
@@ -0,0 +1,189 @@
+		}
+	}
+	...
+}
+```
+
+After Spring 4.3, **it is optional to specify `@Autowired` if the class has only one constructor defined**. However, if we have a class with multiple constructors, we need to explicitly mention `@Autowired` annotation to any one of the constructors for the container to create beans.
+
+# Setter/Field based injection
+
+In setter-based injection, we can provide the required dependencies as field parameters to the class and values are set using the setter methods of the properties. We need to annotate the setter methods with `@Autowired`  annotation. Similarly, in field-based injection, Spring assigns the required dependencies directly to the fields on annotating with `@Autowired` annotation. 
+
+For example, here the `Cake` class requires an object of `Topping` and `Flavor` class. The `Flavor` object is provided as an argument in the setter method of that property and `Topping` object is provided as field value to the Cake class.
+```java
+@Component
+public class Cake {
+
+	private Logger LOGGER = LoggerFactory.getLogger(Cake.class);
+	
+	private Flavor flavor;
+	
+	@Autowired
+	private Topping toppings;
+	
+	public Cake() {
+		LOGGER.info("Flavor from setter Injection : " + this.flavor);
+	}
+	
+	@Autowired
+	public void setFlavor(Flavor flavor) {
+		LOGGER.info("Initialising flavor object using setter injection");
+		this.flavor = flavor;
+	}
+	...
+}
+
+```
+
+# Why should I use constructor injection?
+
+Now that we have seen the different types of injection, let us go through some of the advantages of using constructor injection.
+
+## All required dependencies are available at initialization time
+
+Every class has a default constructor unless we explicitly create a constructor and its object is created by calling respective constructor. If we specify all the required dependencies in the constructor during instance creation, then we can be 100% sure that the class will never be instantiated without its dependencies injected.
+
+The **IoC container makes sure all the arguments provided in the constructor are available before bean creation**. This helps in preventing the infamous `NullPointerException`.
+```java
+@Component
+public class Cake {
+
+	private static Logger LOGGER = LoggerFactory.getLogger(Cake.class);
+
+	private Flavor flavor;
+
+	public Cake(Flavor flavor) throws IllegalAccessException {
+		
+		//check if the required dependency is not null
+		if (flavor != null) {
+			this.flavor = flavor;
+			LOGGER.info("Flavor from Constructor Injection : " + flavor);
+		} else {
+			throw new IllegalArgumentException("Cake cannot be created with null flavor object");
+		}
+	}
+	
+	...
+}
+```
+
+In setter injection, these dependencies are initialized using setter methods only after the constructor invocation. Using setter injection, we can override the existing values of the object and provide different values. This kind of injection is preferable when we have optional dependencies.
```

Comment:
Why is this preferable when we have optional dependencies? Can you explain that in the text?

Reply:
I have added some explanation. Please check and tell if it is okay.